_source/_posts/2023-07-28-scim-workshop.md

```diff
@@ -0,0 +1,1150 @@
+            ...defaultUserSchema,
+            id: id.toString(),
+            userName: updatedUser.email,
+            name: {
+                givenName,
+                familyName
+            },
+            emails: [{
+                primary: true,
+                value: updatedUser.email,
+                type: 'work'
+            }],
+            displayName: updatedUser.name,
+            externalId: updatedUser.externalId,
+            active: updatedUser.active
+        } satisfies IUserSchema;
+    } else if (userCount === 0) {
+        httpStatus = 404;
+        userResponse = `User ${id} not found`;
+    } else {
+        httpStatus = 500;
+        userResponse = `Whoa there!`;
+    }
+    res.status(httpStatus).send(userResponse);
+});
+```
+Does this code make any changes to the incoming data when updating a user? What errors will it catch if the update fails? 
+
+##### Test Updating a user
+
+Let's change our user with ID 3. First, do you remember how to use Postman to look up a user's information when you know their ID? Look up the user with ID 3 now. 
+
+As your GET request to `http://localhost:3333/scim/v2/Users/3` revealed, this person is named Test User. Let's update their `userName`, `givenName`, and primary email to `newTest User`.
+
+To make this update, use Postman to send a PUT request to `http://localhost:3333/scim/v2/Users/3`, with the following body: 
+
+```
+{
+    "schemas": [
+        "urn:ietf:params:scim:schemas:core:2.0:User"
+    ],
+    "locale": "en-US",
+    "groups": [],
+    "meta": {
+        "resourceType": "User"
+    },
+    "id": "3",
+    "userName": "newtest.user@portal.example",
+    "name": {
+        "givenName": "newTest",
+        "familyName": "User"
+    },
+    "emails": [
+        {
+            "primary": true,
+            "value": "newtest.user@portal.example",
+            "type": "work"
+        }
+    ],
+    "displayName": "Test User",
+    "externalId": "00ujl29u0le5T6Aj10h7",
+    "active": true
+}
+```
+
+What response do you expect? You should get an HTTP 200 with a body like this: 
+
+```
+{
+    "schemas": [
+        "urn:ietf:params:scim:schemas:core:2.0:User"
+    ],
+    "locale": "en-US",
+    "groups": [],
+    "meta": {
+        "resourceType": "User"
+    },
+    "id": "3",
+    "userName": "newtest.user@portal.example",
+    "name": {
+        "givenName": "newTest",
+        "familyName": "User"
+    },
+    "emails": [
+        {
+            "primary": true,
+            "value": "newtest.user@portal.example",
+            "type": "work"
+        }
+    ],
+    "displayName": "newTest User",
+    "externalId": "00ujl29u0le5T6Aj10h7",
+    "active": true
+}
+```
+### Support deleting users
+
+Section 3.6 of the  [SCIM spec](https://www.rfc-editor.org/rfc/rfc7644#section-3.6), tells us that service providers may choose to permanently delete users with a 204 (No Content). If no user with the specified ID is found, the server returns a 404 (Not found).
```

Comment:
```suggestion
Section 3.6 of the  [SCIM spec](https://www.rfc-editor.org/rfc/rfc7644#section-3.6), tells us that service providers may choose to permanently delete users with a 204 (No Content). If no user with the specified ID is found, the server returns a 404 (Not found).

When a user leaves an organization, some identity providers will delete that user entirely. Revoking the user's access to all applications is an essential component of securely offboarding them. Additionally, when an enterprise purchases a subscription plan for a SaaS app with a limited number of seats, they expect that removing a user from that app will free up a seat in their license for another user to gain access.

Deleting accounts of removed users can help your application comply with data retention regulations. Over time, storing and backing up information every user who has ever used your app, instead of only tracking active accounts, can get expensive in both storage costs and backup/restore process duration. 

```